### PR TITLE
Likes feature

### PR DESCRIPTION
**Pull request by AlexRS90**
## For this PR @jluisgomezmontal and me:

✅ Implemented button likes feature 
✅ The user is able to give a like to a specific recipe
✅ They can dislike a specific recipe as well
✅ They can see the button in the recipes list and in the show as well
✅ For a better user experience we are using Ajax to display any change automacally
✅ Just sign in users can give likes and just one per recipe

### CODE REVIEW

File: config/database.yml [hunk 1]
```diff
@@ -2,14 +2,14 @@ development:
   adapter: postgresql
   database: myrecipes_development
   username: postgres
-  password: <%= Rails.application.credentials.pg[:db_password] %>
+  password: 220690
```
Comment (coderdave) [inline config/database.yml]: should not be hardcoding your password here. Go back to the previous way the password was in there using credentials. See the Readme file on how to set your credentials.

File: app/assets/stylesheets/likes.scss [hunk 2]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the likes controller here.
```
Comment (coderdave) [inline app/assets/stylesheets/likes.scss]: is this empty file necessary?

File: app/controllers/likes_controller.rb [hunk 3]
```diff
@@ -0,0 +1,16 @@
+class LikesController < ApplicationController
+  before_action :require_user
+
+  def create
+    @like = Like.new(recipe_id: params[:recipe_id])
+    @recipe = Recipe.find(params[:recipe_id])
+    @like.chef = current_chef
+    @like.save
```
Comment (coderdave) [inline app/controllers/likes_controller.rb]: probably good to "bang" the save so it won't fail silently. In the future, we need an enhancement that will flash and error to the user if the save failed.

File: app/helpers/likes_helper.rb [hunk 4]
```diff
@@ -0,0 +1,2 @@
+module LikesHelper
```
Comment (coderdave) [inline app/helpers/likes_helper.rb]: we can remove empty files

File: config/webpack/environment.js [hunk 5]
```diff
@@ -5,8 +5,10 @@ const webpack = require("webpack");
 environment.plugins.append(
   "Provide",
   new webpack.ProvidePlugin({
-    $: "jquery",
-    jQuery: "jquery",
+    $: 'jquery/src/jquery',
```
Comment (coderdave) [inline config/webpack/environment.js:8]: I think all we need is this line. Test it out.

**Review by coderdave — COMMENTED** [review]
why did you delete the cover image?

File: app/controllers/likes_controller.rb [hunk 6]
```diff
@@ -0,0 +1,16 @@
+class LikesController < ApplicationController
+  before_action :require_user
+
+  def create
+    @like = Like.new(recipe_id: params[:recipe_id])
```
Comment (coderdave) [inline app/controllers/likes_controller.rb]: don't need instance variables if they aren't referenced anywhere else

File: app/controllers/likes_controller.rb [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+    @recipe = Recipe.find(params[:recipe_id])
+    @like.chef = current_chef
+    @like.save
+  end
+
+  def destroy
+    @recipe = Recipe.find(params[:recipe_id])
```
Comment (coderdave) [inline app/controllers/likes_controller.rb:14]: don't need instance variables if we aren't using them anywhere else

File: app/assets/images/hand-thumbs-down-fill.svg:Zone.Identifier [hunk 8]
```diff
@@ -0,0 +1,4 @@
+[ZoneTransfer]
```
Comment (coderdave) [inline app/assets/images/hand-thumbs-down-fill.svg:Zone.Identifier]: why are there so many svg files? we need all of them?

File: app/controllers/likes_controller.rb [hunk 9]
```diff
@@ -0,0 +1,16 @@
+class LikesController < ApplicationController
+  before_action :require_user
+
+  def create
+    @like = Like.new(recipe_id: params[:recipe_id])
+    @recipe = Recipe.find(params[:recipe_id])
```
Comment (coderdave) [inline app/controllers/likes_controller.rb:6]: why is there a @recipe var here? Is it being used?

File: app/controllers/likes_controller.rb (hunk 7 above)
Comment (coderdave) [inline app/controllers/likes_controller.rb:14]: why is this @recipe var here? Is it being used?

File: app/controllers/likes_controller.rb (hunk 6 above)
Comment (coderdave) [inline app/controllers/likes_controller.rb]: can also add the chef on this line

Matter of fact, you can condense to one line with `Like.create!`

File: app/controllers/recipes_controller.rb [hunk 10]
```diff
@@ -28,9 +28,9 @@ def create
   end
 
   def edit
-    
+
```
Comment (coderdave) [inline app/controllers/recipes_controller.rb]: best practice in PR's to change as little files as possible when it has nothing to do with your task. I would recommend that you revert this if it doesn't concern this task.

File: app/views/likes/_buttons.html.erb [hunk 11]
```diff
@@ -0,0 +1,20 @@
+<% if logged_in? %>
+  <% if current_chef.likes.find {|like| like.recipe_id == object.id} %>
+    <% like = current_chef.likes.find {|like| like.recipe_id == object.id} %>
+    <%= link_to recipe_like_path(recipe_id: object.id, id: like.id), class: "dislike-button-#{object.id}", remote: true, method: :delete do %>
+      <button class="btn btn-sm active px-3 btn-outline-primary">
+        <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" class="bi bi-hand-thumbs-up-fill" viewBox="0 0 16 16">
```
Comment (coderdave) [inline app/views/likes/_buttons.html.erb]: definitely hard to read these files with the svg embedded in them... Would be nice to use an icon library.. but if we can't, we'll have to fix later...

File: app/views/likes/create.js.erb [hunk 12]
```diff
@@ -0,0 +1,3 @@
+$("#likes_<%= @recipe.id %>").html('<%= @recipe.likes.count %>&nbsp;');
```
Comment (coderdave) [inline app/views/likes/create.js.erb]: I see, you're using the @recipe var in here, so you can resolve the previous comments asking where it is used

Comment (coderdave) [inline app/views/likes/create.js.erb]: even the `@recipe.likes.count` is cached when its called twice, it's best practice to store this in a var so it won't be calculated again.

File: app/views/likes/create.js.erb [hunk 13]
```diff
@@ -0,0 +1,3 @@
...
+$("#like-word_<%= @recipe.id %>").html("<%= @recipe.likes.count == 1 ? 'like' : 'likes' %>");
```
Comment (coderdave) [inline app/views/likes/create.js.erb]: use `pluralize` here from Rails instead of the ternary

File: app/views/likes/destroy.js.erb [hunk 14]
```diff
@@ -0,0 +1,3 @@
+$("#likes_<%= @recipe.id %>").html('<%= @recipe.likes.count %>&nbsp;');
+$("#like-word_<%= @recipe.id %>").html("<%= @recipe.likes.count == 1 ? 'like' : 'likes' %>");
```
Comment (coderdave) [inline app/views/likes/destroy.js.erb]: use pluralize

File: app/views/recipes/_form.html.erb [hunk 15]
```diff
@@ -26,9 +26,8 @@
     <div class="mt-3">
```
Comment (coderdave) [inline app/views/recipes/_form.html.erb:26]: don't change files that doesn't concern your task

File: app/views/recipes/_recipe.html.erb [hunk 16]
```diff
@@ -1,48 +1,39 @@
...
+        <p><%= truncate(recipe.description, length: 150) %></p>
+        <% if recipe.ingredients.any? %>
+          <p>Ingredients: <%= render recipe.ingredients %></p>
+        <% end %>
+        <p>
+          <span class="quiet">
+            <small>
+              Created <%= time_ago_in_words(recipe.created_at) %> ago by <%= recipe.chef.name.capitalize %>
+                | <%= pluralize(recipe.comments.count, 'Comment') %>
+                | <span id="likes_<%= recipe.id %>"><%= recipe.likes.count %></span>
```
Comment (coderdave) [inline app/views/recipes/_recipe.html.erb]: use pluralize

File: db/migrate/20220526132409_create_likes.rb [hunk 17]
```diff
@@ -0,0 +1,10 @@
+class CreateLikes < ActiveRecord::Migration[6.1]
+  def change
+    create_table :likes do |t|
```
Comment (coderdave) [inline db/migrate/20220526132409_create_likes.rb:3]: in the future when naming tables, it's best to be as specific as possible, like `recipe_likes` (but go ahead and leave it for this exercise)

File: package.json [hunk 18]
```diff
@@ -8,6 +8,7 @@
     "@rails/ujs": "^6.0.0",
     "@rails/webpacker": "5.4.3",
     "bootstrap": "^5.1.3",
+    "bootstrap-icons": "^1.8.3",
```
Comment (coderdave) [inline package.json:11]: added but not used? seemed like a good idea? :)

File: test/controllers/likes_controller_test.rb [hunk 19]
```diff
@@ -0,0 +1,24 @@
+# require "test_helper"
+
+# class LikesControllerTest < ActionDispatch::IntegrationTest
```
Comment (coderdave) [inline test/controllers/likes_controller_test.rb]: uh oh! commented out tests! Should make tests when user creates or destroys likes

File: test/models/like_test.rb [hunk 20]
```diff
@@ -0,0 +1,33 @@
+require "test_helper"
```
Comment (coderdave) [inline test/models/like_test.rb:1]: test adding or removing likes

**Review by coderdave — CHANGES_REQUESTED** [review]

File: config/database.yml (hunk 1 above)
Comment (AlexRS90) [reply config/database.yml]: I'll check it out!

Comment (AlexRS90) [reply config/database.yml]: I'll check it out!

File: app/assets/stylesheets/likes.scss (hunk 2 above)
Comment (AlexRS90) [reply app/assets/stylesheets/likes.scss]: nope!
It's already deleted!

File: app/controllers/likes_controller.rb (hunk 3 above)
Comment (AlexRS90) [reply app/controllers/likes_controller.rb]: done!

File: app/helpers/likes_helper.rb (hunk 4 above)
Comment (AlexRS90) [reply app/helpers/likes_helper.rb]: done!

File: config/webpack/environment.js (hunk 5 above)
Comment (AlexRS90) [reply config/webpack/environment.js:8]: Yeah!! is still working!!
Done!

File: app/controllers/likes_controller.rb (hunk 6 above)
Comment (AlexRS90) [reply app/controllers/likes_controller.rb]: I changed the @like variable per a local one!

File: app/assets/images/hand-thumbs-down-fill.svg:Zone.Identifier (hunk 8 above)
Comment (AlexRS90) [reply app/assets/images/hand-thumbs-down-fill.svg:Zone.Identifier]: nope!
SVGs deleted!

File: app/controllers/likes_controller.rb (hunk 9 above)
Comment (AlexRS90) [reply app/controllers/likes_controller.rb:6]: Solved in the nexts comments!

File: app/controllers/likes_controller.rb (hunk 7 above)
Comment (AlexRS90) [reply app/controllers/likes_controller.rb:14]: Solved in the nexts comments!

Comment (AlexRS90) [reply app/controllers/likes_controller.rb:14]: Solved in the below comments!

File: app/controllers/likes_controller.rb (hunk 6 above)
Comment (AlexRS90) [reply app/controllers/likes_controller.rb]: I added a flash message if the like is not saved, I think is a good idea keep Like.new(...)

File: app/controllers/recipes_controller.rb (hunk 10 above)
Comment (AlexRS90) [reply app/controllers/recipes_controller.rb]: Thanks for the feedback we'll have more attention in the next story!

File: app/views/likes/create.js.erb (hunk 12 above)
Comment (AlexRS90) [reply app/views/likes/create.js.erb]: done!

Comment (AlexRS90) [reply app/views/likes/create.js.erb]: Make sence, thanks for the tip I'll do it in future projects
For this one I'll refactor the code to use pluralize and we will call `@recipe.likes.count` just once

File: app/views/recipes/_form.html.erb (hunk 15 above)
Comment (AlexRS90) [reply app/views/recipes/_form.html.erb:26]: Thanks for the feedback, we'll put more attention!

File: db/migrate/20220526132409_create_likes.rb (hunk 17 above)
Comment (AlexRS90) [reply db/migrate/20220526132409_create_likes.rb:3]: make sence, thanks for the feedback!

File: package.json (hunk 18 above)
Comment (AlexRS90) [reply package.json:11]: 🙈 our bad!

File: test/controllers/likes_controller_test.rb (hunk 19 above)
Comment (AlexRS90) [reply test/controllers/likes_controller_test.rb]: working on them!

File: test/models/like_test.rb (hunk 20 above)
Comment (AlexRS90) [reply test/models/like_test.rb:1]: working on them!

File: app/views/likes/create.js.erb (hunk 12 above)
Comment (AlexRS90) [reply app/views/likes/create.js.erb]: done!

File: app/views/likes/destroy.js.erb (hunk 14 above)
Comment (AlexRS90) [reply app/views/likes/destroy.js.erb]: done!

File: app/views/recipes/_recipe.html.erb (hunk 16 above)
Comment (AlexRS90) [reply app/views/recipes/_recipe.html.erb]: done!

File: app/views/likes/create.js.erb (hunk 13 above)
Comment (AlexRS90) [reply app/views/likes/create.js.erb]: done!

File: test/models/like_test.rb (hunk 20 above)
Comment (AlexRS90) [reply test/models/like_test.rb:1]: done!

File: test/controllers/likes_controller_test.rb (hunk 19 above)
Comment (AlexRS90) [reply test/controllers/likes_controller_test.rb]: done!

File: package.json (hunk 18 above)
Comment (AlexRS90) [reply package.json:11]: now we're using this library keeping our code clean and easy to read! 🤓

File: app/views/likes/_buttons.html.erb (hunk 11 above)
Comment (AlexRS90) [reply app/views/likes/_buttons.html.erb]: I agree, is fixed now!

File: config/database.yml (hunk 1 above)
Comment (AlexRS90) [reply config/database.yml]: done!

**Comment by AlexRS90** [comment]
Hi @coderdave, thanks for the feedback and teaching me better ways to code, I think everything is fixed and is working as expected!
Regards,
Alejandro Ramos

**Comment by AlexRS90** [comment]
BTW I don't know when we deleted the cover image, but it's back 😃

**Review by coderdave — CHANGES_REQUESTED** [review]
running `rails test` gives a syntax error. Please fix.

**Pull request merged**